### PR TITLE
WebGLRenderer: Enable rendering without vertex data.

### LINKED ISSUE #19430: InstancedBufferGeometry with no used attributes not rendered
##### Description of the problem
Rendering an InstancedBufferGeometry with no (used) attributes does not render.

This happens when a shader uses only (implicit) gl_InstanceID and/or gl_VertexID,
and user program establishes count with 'geom.maxInstancedCount = ...'
(It only applies to WebGL2 as earlier shaders do not support this.)

The internal value geometry._maxInstanceCount is never established, so 
var instanceCount = Math.min( geometry.instanceCount, geometry._maxInstanceCount );
does not set a sensible value.

A workaround is to set geometry._maxInstanceCount  in the user's program (eg to Infinity),
but this is obviously not the correct solution.

##
With earlier versions of three.js (eg 109) a program could provide a 'dummy' attribute not actually used in the shader to establish count; but 117 is clever enough to realize this dummy is not used, and therefore does not take account of it in the count calculation.

##
example to follow

##### Three.js version

- [ ] Rel
- [ ] r117
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### PR DESCRIPTION
**Pull request by filonik**
Fixed https://github.com/mrdoob/three.js/issues/19430

**Description**

Attempt at cleaning up pull request https://github.com/mrdoob/three.js/pull/19451 to make the logic/control flow easier to follow. The goal is to enable rendering procedural geometries without vertex data (using raw shader materials).

Implements the following logic:

1. Initialize `drawStart` and `drawEnd` to `0` and `Infinity`.
2. If a `drawRange` is specified, clamp `drawStart` and `drawEnd` to `drawRange.start` and `drawRange.end`.
3. If a `group` is specified, clamp `drawStart` and `drawEnd` to `group.start` and `group.end`.
4. If the geometry is indexed, clamp `drawStart` and `drawEnd` to `0` and `index.count`.
5. If the geometry is non-indexed with a position attribute, clamp `drawStart` and `drawEnd` to `0` and `position.count` .

### CODE REVIEW

File: src/renderers/WebGLRenderer.js
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js:756]: Please avoid the creation of objects in functions/methods which are invocated per frame multiple times. Object creation (even inline) is an expensive operation and should not be done over and over again.

Comment (filonik) [reply src/renderers/WebGLRenderer.js:756]: Fair point. I put my faith in modern JavaScript engines to optimize that sort of thing, but it's probably better not to rely on it. I will update the code.

**Comment by filonik** [comment]
I created an example with a simple usage example of a geometry without attributes. Is there any documentation on how to add a description/screenshot? (CI fails because of it...)

File: src/renderers/WebGLRenderer.js
Comment (LeviPesin) [inline src/renderers/WebGLRenderer.js:767]: Why not just `if ( drawCount < 0 )`? I think infinite `drawCount` is allowed?

Comment (filonik) [reply src/renderers/WebGLRenderer.js:767]: I am not sure if it is, since this value is basically passed directly to `gl.drawArrays`/`gl.drawElements`.

https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/drawArrays

**Comment by filonik** [comment]
> Via the `npm run make-screenshot examplename`.

I tried this, and I am getting:

> ERROR! Diff wrong in 0.095 of pixels in file: webgl2_buffergeometry_no_attributes

Maybe this is platform dependent?

File: src/renderers/WebGLRenderer.js
Comment (filonik) [reply src/renderers/WebGLRenderer.js:767]: I just did a quick test, and somewhat to my surprise, passing `Infinity` doesn't seem to cause any issues (aside from nothing being rendered obviously). However, I am still not sure if it is a good idea though. Is it specified somewhere that in infinite count is valid for `gl.drawArrays`/`gl.drawElements`?

Comment (filonik) [reply src/renderers/WebGLRenderer.js:767]: Basically, my reasoning is that if the `drawCount` is still `Infinity` at this point, it means that it wasn't set explicitly in `drawRange` or `group`, or implicitly by `index` or a `position` attribute. In that case, I would consider the geometry to be empty/incomplete.

**Comment by filonik** [comment]
> > Via the `npm run make-screenshot examplename`.
> 
> I tried this, and I am getting:
> 
> > ERROR! Diff wrong in 0.095 of pixels in file: webgl2_buffergeometry_no_attributes
> 
> Maybe this is platform dependent? I use some pseudo random noise in the example. It should be deterministic, but maybe the results differ between platforms.
> 
> I can move the example to a separate pull request if that would be preferable.

Ok, I swapped out the noise function for a slightly more complicated but stable one and CI finally seems happy.

File: src/renderers/WebGLRenderer.js
Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:767]: That reasoning makes sense to me. Still, could you write the check like so:
```js
if ( drawCount < 0 || drawCount === Infinity ) return;
```

**Comment by Mugen87** [comment]
This is a real nice refactoring! 👍

File: src/renderers/WebGLRenderer.js
Comment (filonik) [reply src/renderers/WebGLRenderer.js]: Sure, I have no strong preference for how the check is written. Your version does look a bit cleaner, so I've adopted it.

**Review by Mugen87 — DISMISSED** [review]
Okay, let's try this!

**Comment by Mugen87** [comment]
I have noticed one thing after the approval. Is the use case mentioned in #19430 actually fixed by this PR?

It works now if you define a draw range on `BufferGeometry` level but not when using `InstancedBufferGeometry`. In this case, you only define `instanceCount` and no draw range. I guess we have to remove the `drawCount === Infinity` check otherwise the following piece of code is never reached. 

https://github.com/mrdoob/three.js/blob/76c7c1c3ea0078f0b29400e0ba6a4894845315a3/src/renderers/WebGLRenderer.js#L831-L833

**Comment by filonik** [comment]
Hm, I would have thought that `drawCount` essentially works the same way for instanced geometries. I am still not sure what the expected result passing `Infinity` actually is. I would have thought that both  `drawCount` and `instanceCount` need to be some positive integer in order for something to be rendered. Let me investigate further...

File: src/renderers/WebGLRenderer.js
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js:839]: I don't think we should touch this code block in this PR.

Your example assumes you use `setDrawRange()` with `InstancedBufferGeometry` which should not be necessary. I think we should keep the assumption that `Infinity` means "render everything in the buffer".

Comment (filonik) [reply src/renderers/WebGLRenderer.js:839]: I can remove the check for `Infinity`, but it won't fix the problem in #19430, since `_maxInstanceCount` will still be `undefined`.

Comment (filonik) [reply src/renderers/WebGLRenderer.js:839]: I am not sure I understand the use case for an `InstancedBufferGeometry` with `drawCount` set to `Infinity`.

https://developer.mozilla.org/en-US/docs/Web/API/WebGL2RenderingContext/drawArraysInstanced

Even when you are rendering multiple instances, you need to somehow specify what range of vertices you want to render multiple instances of. The correct way of handling #19430 is to specify a vertex count (`drawRange.count`) and a instance count. Maybe there's some use case that I am not seeing...

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:839]: I had to debug this issue more closely to understand what's going on  😅 .

`_maxInstanceCount` is `undefined` in your example since `setupVertexAttributes()` in `WebGLBindingStates` never sets it. That makes sense however since the geometry has no attributes. So you are right, the additional check is actually required.

I've also verified that `Infinity` is not a valid input for all WebGL drawing methods. So it's true that we can't use it as an input.

I'm feeling more confident with the PR now. The only thing that can be changed is the beginning of the change:

```js
const drawRange = geometry.drawRange;
const position = geometry.attributes.position;

let drawStart = drawRange.start * rangeFactor ;
let drawEnd = ( drawRange.start + drawRange.count ) * rangeFactor;
```
We can use this simplification since `drawRange` is always defined when using `BufferGeometry`. It is not a nullable property. So the respective `if` clause can be removed.

Comment (filonik) [reply src/renderers/WebGLRenderer.js:839]: Yes, this reflects my understanding of things. Initializing `drawStart`/`drawEnd` as you suggest makes sense. I suppose in this instance I was being overly defensive, trying to account for the possibility of `drawRange` being set to `null`. If this will never happen your approach makes more sense, I will adopt it.

Otherwise, I have fairly high confidence that the current code is correct, and that it is a simplification/improvement over the existing code. The only possible alternative that I see for checking `_maxInstanceCount` for `undefined` is to initialize it to `Infinity` in `setupVertexAttributes`. However, I am not sure if that would really be an improvement.

**Review by Mugen87 — APPROVED** [review]

**Pull request merged**
